File: python/data_structures/linked_lists/1_traverse_and_print.py
"""
Exercise: Traverse and Print (Easy)

Given the head of a singly linked list, traverse the list and return all values in a list.

Example:
Input: 1 -> 2 -> 3 -> 4 -> None
Output: [1, 2, 3, 4]

Time Complexity: O(n) where n is the number of nodes
Space Complexity: O(n) for the output list
"""

from typing import List, Optional
from .list_node import ListNode


def traverse(head: Optional[ListNode]) -> List[int]:
    """
    Traverse the linked list and return all values.

    Args:
        head: The head node of the linked list

    Returns:
        A list containing all values from the linked list
    """
    # TODO: Implement this function
    # Hint: Start at head and follow next pointers until you reach None
    # Add each value to a list
    raise NotImplementedError("Not implemented yet")


if __name__ == "__main__":
    # Example usage
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
    print(f"Values: {traverse(head)}")

File: rust/data-structures/src/linked_lists/1_traverse_and_print.rs
/// Exercise: Traverse and Print (Easy)
///
/// Given the head of a singly linked list, traverse the list and return all values in a vector.
///
/// Example:
/// Input: 1 -> 2 -> 3 -> 4 -> None
/// Output: vec![1, 2, 3, 4]
///
/// Time Complexity: O(n) where n is the number of nodes
/// Space Complexity: O(n) for the output vector

use super::list_node::ListNode;

pub fn traverse(head: Option<Box<ListNode>>) -> Vec<i32> {
    // TODO: Implement this function
    // Hint: Use pattern matching with Option to traverse until you reach None
    // Add each value to a vector
    unimplemented!("Not implemented yet")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_list() {
        assert_eq!(traverse(None), vec![]);
    }

    #[test]
    fn test_single_node() {
        let head = Some(Box::new(ListNode::new(42)));
        assert_eq!(traverse(head), vec![42]);
    }

    #[test]
    fn test_multiple_nodes() {
        let head = Some(Box::new(ListNode::with_next(
            1,
            Some(Box::new(ListNode::with_next(
                2,
                Some(Box::new(ListNode::with_next(
                    3,
                    Some(Box::new(ListNode::new(4))),
                ))),
            ))),
        )));
        assert_eq!(traverse(head), vec![1, 2, 3, 4]);
    }

    #[test]
    fn test_two_nodes() {
        let head = Some(Box::new(ListNode::with_next(
            10,
            Some(Box::new(ListNode::new(20))),
        )));
        assert_eq!(traverse(head), vec![10, 20]);
    }

    #[test]
    fn test_negative_values() {
        let head = Some(Box::new(ListNode::with_next(
            -5,
            Some(Box::new(ListNode::with_next(
                -10,
                Some(Box::new(ListNode::new(-15))),
            ))),
        )));
        assert_eq!(traverse(head), vec![-5, -10, -15]);
    }
}
